### PR TITLE
Module aliases override updates

### PR DESCRIPTION
**Pull request by elsh**
- Use alias value defined upstream as a key downstream for overriding and chain them
- Add pre-chained aliases to targets when needed and show diags
Resolves rdar://93445417, rdar://93218209

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/PackageGraph/ModuleAliasTracker.swift [hunk 1]
```diff
@@ -221,20 +254,38 @@ class ModuleAliasTracker {
...
-                                // If there are aliases, remove as aliasing should
-                                // not be applied
+
+                            if !targetsWithConflictingAliases.isEmpty {
+                                // If there are conflicting aliases, remove them
+                                // and show a diag message
+                                let conflictingAliases = targetsWithConflictingAliases.compactMap { $0.moduleAliases?[nameToBeAliased] }
+                                var message = conflictingAliases.isEmpty ? "Targets named '\(nameToBeAliased)' found in other dependency products;" : "Conflicting module aliases found \(conflictingAliases + [aliasInCurChild]) for '\(nameToBeAliased)'; "
+                                message += "these will be removed from target '\(curTarget.name)'"
+                                observabilityScope.emit(info: message)
```
Comment (tomerd) [inline Sources/PackageGraph/ModuleAliasTracker.swift:271]: is this informational, or should be a warning?

File: Tests/BuildTests/ModuleAliasingBuildTests.swift [hunk 2]
```diff
@@ -1146,13 +1146,12 @@ final class ModuleAliasingBuildTests: XCTestCase {
...
         XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "FooLogging" && $0.target.moduleAliases?["Logging"] == "FooLogging" })
-        XCTAssertFalse(result.targetMap.values.contains { $0.target.name == "BarLogging" && $0.target.moduleAliases?["Logging"] == "BarLogging" })
         XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "FooMath" && $0.target.moduleAliases?["Math"] == "FooMath" })
-        XCTAssertFalse(result.targetMap.values.contains { $0.target.name == "BarMath" && $0.target.moduleAliases?["Math"] == "BarMath" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases?["Logging"] == "FooLogging" && $0.target.moduleAliases?["Math"] == "FooMath" })
         XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Logging" && $0.target.moduleAliases == nil })
         XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Math" && $0.target.moduleAliases == nil })
+        XCTAssertFalse(result.targetMap.values.contains { $0.target.name == "BarLogging" })
```
Comment (tomerd) [inline Tests/BuildTests/ModuleAliasingBuildTests.swift:1154]: are changes to existing tests expected because they were logically wrong before? not sure if this is a problem in this case, but its always a bit concerning to see changes to how existing tests treats output when the inputs are the same / similar

**Review by tomerd — APPROVED** [review]
a bit hard to follow all the changes, but adding new tests is great

**Comment by elsh** [comment]
@swift-ci  smoke test

File: Tests/BuildTests/ModuleAliasingBuildTests.swift (hunk 2 above)
Comment (elsh) [reply Tests/BuildTests/ModuleAliasingBuildTests.swift:1155]: This is an additive check, which was skipped previously.

File: Sources/PackageGraph/ModuleAliasTracker.swift (hunk 1 above)
Comment (elsh) [reply Sources/PackageGraph/ModuleAliasTracker.swift]: this is info; changed wording

**Comment by elsh** [comment]
@swift-ci smoke test
